src/expr/core/src/window_function/state/buffer.rs

```diff
@@ -258,19 +172,184 @@ impl<K: Ord, V: Clone> WindowBuffer<K, V> {
             .drain(0..min_needed_idx)
             .map(|Entry { key, value }| (key, value))
     }
+
+    fn maintain_delta(&mut self, old_outer: Range<usize>, new_outer: Range<usize>) {
+        debug_assert!(self.frame_exclusion.is_no_others());
+
+        let (outer_removed, outer_added) = range_diff(old_outer.clone(), new_outer.clone());
+        let delta = self.curr_delta.as_mut().unwrap();
+        for idx in outer_removed.iter().cloned().flatten() {
+            delta.push((Op::Delete, self.buffer[idx].value.clone()));
+        }
+        for idx in outer_added.iter().cloned().flatten() {
+            delta.push((Op::Insert, self.buffer[idx].value.clone()));
+        }
+    }
+
+    fn recalculate_left_right(&mut self) {
+        let buffer_ref = BufferRefMut {
+            buffer: &self.buffer,
+            curr_idx: &mut self.curr_idx,
+            left_idx: &mut self.left_idx,
+            right_excl_idx: &mut self.right_excl_idx,
+        };
+        self.window_impl.recalculate_left_right(buffer_ref);
+    }
+}
+
+/// Wraps a reference to the buffer and some indices, to be used by [`WindowImpl`]s.
+#[derive(Educe)]
+#[educe(Clone, Copy)]
+pub(super) struct BufferRef<'a, K: Ord, V: Clone> {
+    buffer: &'a VecDeque<Entry<K, V>>,
+    curr_idx: usize,
+    left_idx: usize,
+    right_excl_idx: usize,
+}
+
+/// Wraps a reference to the buffer and some mutable indices, to be used by [`WindowImpl`]s.
+pub(super) struct BufferRefMut<'a, K: Ord, V: Clone> {
+    buffer: &'a VecDeque<Entry<K, V>>,
+    curr_idx: &'a mut usize,
+    left_idx: &'a mut usize,
+    right_excl_idx: &'a mut usize,
+}
+
+/// A trait for sliding window implementations. This trait is used by [`WindowBuffer`] to
+/// determine the status of current window and how to slide the window.
+pub(super) trait WindowImpl {
+    type Key: Ord;
+    type Value: Clone;
+
+    /// Whether the preceding half of the current window is saturated.
+    /// By "saturated" we mean that every row that is possible to be in the preceding half of the
```

Comment:
Feel confused from my understanding of the English word "saturated" 😕

How about "buffered", "loaded" or "cached"?

Reply:
Maybe we can change the word later